File: lib/ts/src/internal/validation/ValidateStructFields.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { ValidationFailure } from '../../internal/validation/ValidationFailure.js';
import { TFieldDeclaration } from '../types/TFieldDeclaration.js';
import { ValidateContext } from './ValidateContext.js';

export function validateStructFields(
    fields: Record<string, TFieldDeclaration>,
    selectedFields: string[] | null,
    actualStruct: Record<string, any>,
    ctx: ValidateContext,
): ValidationFailure[] {
    const validationFailures: ValidationFailure[] = [];

    const missingFields: string[] = [];
    for (const [fieldName, fieldDeclaration] of Object.entries(fields)) {
        const isOptional = fieldDeclaration.optional;
        const isOmittedBySelect = selectedFields !== null && !selectedFields.includes(fieldName);
        if (!(fieldName in actualStruct) && !isOptional && !isOmittedBySelect) {
            missingFields.push(fieldName);
        }
    }

    for (const missingField of missingFields) {
        const validationFailure = new ValidationFailure([], 'RequiredObjectKeyMissing', {
            key: missingField,
        });

        validationFailures.push(validationFailure);
    }

    for (const [fieldName, fieldValue] of Object.entries(actualStruct)) {
        const referenceField = fields[fieldName];
        if (referenceField === undefined) {
            const validationFailure = new ValidationFailure([fieldName], 'ObjectKeyDisallowed', {});

            validationFailures.push(validationFailure);
            continue;
        }

        const refFieldTypeDeclaration = referenceField.typeDeclaration;

        ctx.path.push(fieldName);

        const nestedValidationFailures = refFieldTypeDeclaration.validate(fieldValue, ctx);

        ctx.path.pop();

        const nestedValidationFailuresWithPath: ValidationFailure[] = [];
        for (const failure of nestedValidationFailures) {
            const thisPath = [fieldName, ...failure.path];

            nestedValidationFailuresWithPath.push(new ValidationFailure(thisPath, failure.reason, failure.data));
        }

        validationFailures.push(...nestedValidationFailuresWithPath);
    }

    return validationFailures;
}

File: lib/java/src/main/java/io/github/telepact/internal/validation/ValidateStructFields.java
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package io.github.telepact.internal.validation;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.github.telepact.internal.types.TFieldDeclaration;
import io.github.telepact.internal.types.TTypeDeclaration;

public class ValidateStructFields {
    static List<ValidationFailure> validateStructFields(
            Map<String, TFieldDeclaration> fields,
            List<String> selectedFields,
            Map<String, Object> actualStruct, ValidateContext ctx) {
        final var validationFailures = new ArrayList<ValidationFailure>();

        final var missingFields = new ArrayList<String>();
        for (final var entry : fields.entrySet()) {
            final var fieldName = entry.getKey();
            final var fieldDeclaration = entry.getValue();
            final boolean isOptional = fieldDeclaration.optional;
            final boolean isOmittedBySelect = selectedFields != null && !selectedFields.contains(fieldName);
            if (!actualStruct.containsKey(fieldName) && !isOptional && !isOmittedBySelect) {
                missingFields.add(fieldName);
            }
        }

        for (final var missingField : missingFields) {
            final var validationFailure = new ValidationFailure(List.of(),
                    "RequiredObjectKeyMissing",
                    Map.of("key", missingField));

            validationFailures.add(validationFailure);
        }

        for (final var entry : actualStruct.entrySet()) {
            final var fieldName = entry.getKey();
            final var fieldValue = entry.getValue();

            final var referenceField = fields.get(fieldName);
            if (referenceField == null) {
                var validationFailure = new ValidationFailure(List.of(fieldName), "ObjectKeyDisallowed",
                        Map.of());

                validationFailures.add(validationFailure);
                continue;
            }

            final TTypeDeclaration refFieldTypeDeclaration = referenceField.typeDeclaration;

            ctx.path.push(fieldName);

            final var nestedValidationFailures = refFieldTypeDeclaration.validate(fieldValue, ctx);

            ctx.path.pop();

            final var nestedValidationFailuresWithPath = new ArrayList<ValidationFailure>();
            for (final var f : nestedValidationFailures) {
                final List<Object> thisPath = new ArrayList<>(f.path);
                thisPath.add(0, fieldName);

                nestedValidationFailuresWithPath.add(new ValidationFailure(thisPath, f.reason, f.data));
            }

            validationFailures.addAll(nestedValidationFailuresWithPath);
        }

        return validationFailures;
    }
}
